libsolidity/analysis/DeclarationTypeChecker.cpp

```diff
@@ -78,26 +78,61 @@ bool DeclarationTypeChecker::visit(StructDefinition const& _struct)
 	}
 
 	bool previousRecursiveStructSeen = m_recursiveStructSeen;
+	bool previousHasNestedMapping = m_nestedMappingSeen;
 	bool hasRecursiveChild = false;
+	bool hasNestedMapping = false;
 
 	m_currentStructsSeen.insert(&_struct);
 
 	for (auto const& member: _struct.members())
 	{
 		m_recursiveStructSeen = false;
+		m_nestedMappingSeen = false;
 		member->accept(*this);
 		solAssert(member->annotation().type, "");
 		solAssert(member->annotation().type->canBeStored(), "Type cannot be used in struct.");
+
+		if (member->type()->category() == Type::Category::Mapping)
+			hasNestedMapping = true;
+		else if (auto arrayType = dynamic_cast<ArrayType const*>(member->type()))
+		{
+			TypePointer finalBaseType = arrayType->finalBaseType(false);
+			if (finalBaseType->category() == Type::Category::Mapping)
+				hasNestedMapping = true;
+			else if (auto structType = dynamic_cast<StructType const*>(finalBaseType))
+			{
+				optional<bool> containsNestedMapping =
+					structType->structDefinition().annotation().containsNestedMapping;
+				if (containsNestedMapping.has_value() && containsNestedMapping.value())
+					hasNestedMapping = true;
+			}
+		}
+		else if (auto structType = dynamic_cast<StructType const*>(member->type()))
+		{
+			optional<bool> containsNestedMapping =
+				structType->structDefinition().annotation().containsNestedMapping;
+			if (containsNestedMapping.has_value() && containsNestedMapping.value())
```

Comment:
Shouldn't this be
```suggestion
			solAssert(containsNestedMapping.has_value(), "");
			if (containsNestedMapping.value())
```

Reply:
That would cause problems for direct recursive structs.

```solidity
	  struct S {T t;}
    struct T {S s;}
```

Instead of an error (line 160) we'll now have an ICE.